### DISCUSSION TITLE
Remove &nbsp; by default from Latex-Code before processing by Katex?

### DISCUSSION THREAD

**Discussion by q2apro** [KaTeX/KaTeX#2606] [Q&A] [answered]
I am not sure if I should post it as bug or feature request, so "general":

We have noticed that our editor (also found in other editors like Gmail), sets sometimes `&nbsp;` as whitespace, which is not visible frontend (wysiwyg).

Then, Katex preserves this whitespace and the rendering gets an extra whitespace.

![image](https://user-images.githubusercontent.com/6764548/97068330-1fbfa900-15cf-11eb-9bc5-b17b5eeabed6.png)

See the last `2` which has one whitespace before.

Would it make sense to have Katex (we are using the autorenderer!) remove the `&nbsp;` before rendering (or better replace them with real whitespaces)?

Is this already possible with the current version? If yes, how to tell Katex to do that?

**Comment by edemaine** [comment]
I believe you could fix this with the `preProcess` option of autorender. See [the documentation](https://katex.org/docs/autorender.html). Something like `preProcess: (math) => math.replace(/&nbsp;/g, ' ')` (or you may need to adjust the search string to be an actual unicode character, or `&#` escape).

**Comment by q2apro** [comment]
Thanks for your reply. 

I tried:

	renderMathInElement(document.body, {
		delimiters: [
			{left: "$$", right: "$$", display: true},
			{left: "\\(", right: "\\)", display: false},
		],
		fleqn: true,
		preProcess: (math) => math.replace(/&nbsp;/g, ' '),  // NEW
	});

but the preProcess seems not to change anything. Maybe the syntax at `preProcess` is incorrect?

**Comment by q2apro** [comment]
The problem still persists. If anyone knows a solution or workaround, feel free to let us know. Thanks!

**Comment by edemaine** [comment, marked as answer]
It sounds like you're getting the actual nonbreaking space character, not the string `&nbsp;'` in your content.  Try `preProcess: (math) => math.replace(/\xa0/g, ' ')`

**Reply by q2apro** [reply]
It works now. Thanks a lot. 👍
